### ISSUE TITLE
Data race in buffer with v2.4.3

### ISSUE THREAD

**Issue by jkager** [mochi-mqtt/server#346]
I keep getting data races with v2.4.3 while running tests in GitHub workflows:

```
WARNING: DATA RACE
Write at 0x00c00137f0e0 by goroutine 1116:
  bytes.(*Buffer).grow()
      /opt/hostedtoolcache/go/1.21.4/x64/src/bytes/buffer.go:151 +0x2d3
  bytes.(*Buffer).Write()
      /opt/hostedtoolcache/go/1.21.4/x64/src/bytes/buffer.go:179 +0xc4
  bytes.(*Buffer).WriteTo()
      /opt/hostedtoolcache/go/1.21.4/x64/src/bytes/buffer.go:261 +0x106
  github.com/mochi-mqtt/server/v2.(*Client).WritePacket.func1()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/clients.go:609 +0x2c4
  github.com/mochi-mqtt/server/v2.(*Client).WritePacket()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/clients.go:616 +0x6b9
  github.com/mochi-mqtt/server/v2.(*Client).WriteLoop()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/clients.go:198 +0x1c4
  github.com/mochi-mqtt/server/v2.(*Server).NewClient.func1()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/server.go:237 +0x33

Previous read at 0x00c00137f0e0 by goroutine 1115:
  bytes.(*Buffer).Bytes()
      /opt/hostedtoolcache/go/1.21.4/x64/src/bytes/buffer.go:54 +0x804
  github.com/mochi-mqtt/server/v2.(*Client).WritePacket()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/clients.go:627 +0x848
  github.com/mochi-mqtt/server/v2.(*Server).processPubrec()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/server.go:1089 +0x3a4
  github.com/mochi-mqtt/server/v2.(*Server).processPacket()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/server.go:601 +0x2a7
  github.com/mochi-mqtt/server/v2.(*Server).receivePacket()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/server.go:435 +0x6c
  github.com/mochi-mqtt/server/v2.(*Server).receivePacket-fm()
      <autogenerated>:1 +0x87
  github.com/mochi-mqtt/server/v2.(*Client).Read()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/clients.go:381 +0x1bc
  github.com/mochi-mqtt/server/v2.(*Server).attachClient()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/server.go:390 +0xa45
  github.com/mochi-mqtt/server/v2.(*Server).EstablishConnection()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/server.go:327 +0x7e
  github.com/mochi-mqtt/server/v2.(*Server).EstablishConnection-fm()
      <autogenerated>:1 +0x64
  github.com/mochi-mqtt/server/v2/listeners.(*Net).Serve.func1()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/listeners/net.go:68 +0x91

Goroutine 1116 (running) created at:
  github.com/mochi-mqtt/server/v2.(*Server).NewClient()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/server.go:237 +0x385
  github.com/mochi-mqtt/server/v2.(*Server).EstablishConnection()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/server.go:326 +0x67
  github.com/mochi-mqtt/server/v2.(*Server).EstablishConnection-fm()
      <autogenerated>:1 +0x64
  github.com/mochi-mqtt/server/v2/listeners.(*Net).Serve.func1()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/listeners/net.go:68 +0x91

Goroutine 1115 (running) created at:
  github.com/mochi-mqtt/server/v2/listeners.(*Net).Serve()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/listeners/net.go:67 +0x226
  github.com/mochi-mqtt/server/v2/listeners.(*Listeners).Serve.func1()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/listeners/listeners.go:89 +0x43
  github.com/mochi-mqtt/server/v2/listeners.(*Listeners).Serve.func3()
      /home/runner/go/pkg/mod/github.com/mochi-mqtt/server/v2@v2.4.3/listeners/listeners.go:90 +0x41
```

**Comment by thedevop** [comment]
Thanks @jkager. Found the issue, will issue a fix soon.

**Comment by mochi-co** [comment]
@jkager This should be fixed in v2.4.4 - please give it a try and let us know.

**Comment by thedevop** [comment]
@jkager , can you please verify the data race issue is addressed and if we can close this issue?

**Comment by thedevop** [comment]
This issue should be fixed in v2.4.4. @jkager , please reopen if you continue to see issue.
